### ISSUE TITLE
Focus gets reset on every datagrid data change (add row, remove row, etc.)

### ISSUE THREAD

**Issue by immihaylova** [vmware/clarity#4684]
## Describe the bug

This is an accessibility issue.
We got a report from AccessLevel about it in our product (SRM UI).

We have a clarity datagrid with data that is refreshed every 5 seconds. If the user uses the keyboard to navigate through the datagrid rows, the focus will get reset every 5 seconds, even in the cases when the rows data has not changed visually (though we request the data, receive it again from the server, and set it to the rows items binding).
We also make requests each time a row in the grid gets expanded/collapsed, so each time the user does this through the keyboard, the focus is again moved to the datagrid itself.

Looking at the datagrid code, it seems that each time anything in the clarity datagrid items data changes, the focus gets reset to the datagrid itself.
I couldn't find a mechanism in the datagrid api that let us focus back the item that was previously focused.

## How to reproduce

1.  Add a clarity datagrid with 1-2 rows in a page, and update the data (push new item, or delete an item, or simply set the items to the same values)  on any time interval convenient.
2. Focus with the keyboard any row (using tab), and wait for the items to update => Each time the items update (even with the same values), the focus gets reset to the datagrid itself.

## Expected behavior

Focus remains unchanged.
The accessibility team from AccessLevel is concerned that focus is changed without user initiating it.
This looks like an issue for live views that get updates often. In these cases, the focus will jump back to the datagrid on each update.

## Versions

**App**

- Angular: [9]
- Node: [10]
- Clarity: [3]

**Device:**

- Type: [Desktop]
- OS: [Win10]
- Browser [chrome]
- Version [81]

**Comment by immihaylova** [comment]
I created a stackblitz here: https://clarity-v3-light-theme-no7afv.stackblitz.io

**Comment by hippee-lee** [comment]
Adding a trackBy fn prevents the rows from being recreated in the DOM. Here is a modified version of the stackblitz: https://stackblitz.com/edit/clarity-v3-light-theme-vp1wbj

I think this will solve your issue but if not please feel free to re-open and add more info.

**Comment by immihaylova** [comment]
Hi Matt,

Thanks for the stackblitz!
It really works fine when nothing changes at all. But the focus is still reset if a row is added (although the focused item is still present in the datagrid). 
I have updated the stackblitz to illustrate that - 5 times, new rows get added, then they just get reloaded. So, while new rows are being added, the focus gets reset, and when the rows only get reloaded without any changes, the focus remains unchanged.

https://stackblitz.com/edit/clarity-v3-light-theme-vp1wbj

Help on solving this would be much appreciated. Thanks.

**Comment by hippee-lee** [comment]
Hi @immihaylova - I think you need to fork my example in order to save it with changes. 

If I understand what you are saying though, even though the first two rows are the same the new, additional row triggers an internal rendering cycle for / with the new row. One suggestion that comes to mind is to track focus on the application side and then put it back on the focused row after the update. 

@clane - can you give me some more insight into guidlines / specs around focused row in datagrid when the datagrid content changes?

**Comment by immihaylova** [comment]
Sorry @hippee-lee, I did forget to fork :)
Here is the new link: https://stackblitz.com/edit/clarity-v3-light-theme-zrqtyd

**Comment by hippee-lee** [comment]
Thanks for the link. Currently rows (and cells) don't have a method to set focus on them and I don't think we can set focus on the clr-dg-row element. 

@chlane - can you weigh in here wrt the use case:
1) datagrid row has focus
2) datagrid model is refreshed every 5 seconds
3) when there are new rows we redraw the rows and update the DOM 
4) users focus is lost

Whats the guideline here? Currently datagrid is setup to enable focus but we have no way to internalize what is focused and reset it.

**Comment by chlane** [comment]
We should provide arrow key navigation between cells. This is a core requirement of an ARIA grid as specified at https://www.w3.org/TR/wai-aria-practices-1.1/#grid. This was requested in 
https://github.com/vmware/clarity/issues/3349 but that issue was closed without addressing this problem. Once we have arrow key navigation, we should track the last grid cell and focus it on every refresh.

**Comment by chlane** [comment]
We should also provide some visible or screen reader only text that will tell assistive technology users that the grid will be refreshing every 5 seconds

**Comment by mathisscott** [comment]
> We should provide arrow key navigation between cells. This is a core requirement of an ARIA grid as specified at https://www.w3.org/TR/wai-aria-practices-1.1/#grid. This was requested in
> #3349 but that issue was closed without addressing this problem. Once we have arrow key navigation, we should track the last grid cell and focus it on every refresh.

Problem: there's no guarantee the last grid cell will be there. With @chlane recommendation, we could focus on the wrong cell or try to focus on a cell that no longer exists.

**Comment by mathisscott** [comment]
> We should also provide some visible or screen reader only text that will tell assistive technology users that the grid will be refreshing every 5 seconds

@chlane 
The every-5-second thing is not a Clarity issue. This is an implementation issue where a product is refreshing the datagrid every 5 seconds.

We need to make sure we are clear on those kinds of details before making recommendations of work for the component library.

@gnomeontherun

**Comment by gnomeontherun** [comment]
Hi @immihaylova, product specific issues should go through the Accessibility team and not through Clarity. I'm going to close this as out of scope for now and to help follow our internal guidance on traceability of issues. Please work with your accessibility contact for next steps.

The recommendations are beyond what Clarity can provide out of the box and a use case we need to correctly define before making further recommendations.

**Comment by immihaylova** [comment]
> Hi @immihaylova, product specific issues should go through the Accessibility team and not through Clarity. I'm going to close this as out of scope for now and to help follow our internal guidance on traceability of issues. Please work with your accessibility contact for next steps.

Hi @gnomeontherun, I don't understand how resetting the focus on each data change is a product specific issue - the code that does this is inside the clarity datagrid, and there is no way to prevent or change this behavior. There are several stackblitz links to illustrate the issue.

I also don't understand what you mean by asking me to work with our accessibility contact. SRM is a VMware product, and got a report from AccessLevel, and they are the ones that filed the issue as a top severity problem with accessibility. We don't have an accessibility person in our VMware team.

A mechanism through which products that use clarity datagrid can manage the focus themselves is also an option.

**Comment by jordanovi** [comment]
Hi team

Our request here is:
1. Provide arrow key navigation between cells. As mentioned above, this is a core requirement of an ARIA grid as specified at https://www.w3.org/TR/wai-aria-practices-1.1/#grid. 
2. Once we have arrow key navigation, track the last grid cell and focus it on every refresh.

Do you want us to file another issue for the request above or do you prefer to track it here (and re-open this issue) ?

**Comment by gnomeontherun** [comment]
Please see https://github.com/vmware/clarity/issues/2356 for key arrow navigation issue.

**Comment by gnomeontherun** [comment]
@immihaylova Feel free to reach out to me over Slack/email, as I do not know for sure who I'm talking to, and I can help explain further.

**Comment by chlane** [comment]
I'm updating my recommendation since resetting focus every 5 seconds will be too disruptive for screen reader users. 

I recommend to remove the auto-refresh behavior and provide a button to allow the user to initiate each refresh.

**Comment by chlane** [comment]
After an email conversation with Stefan Marinov, I recommended using aria-live=”polite” to announce the updates. I also recommend an implementation that decouples these live region announcements from the focus management used  in https://github.com/vmware/clarity/issues/3370.  Also to prevent overly frequent live region announcements, I recommend implementing a button or set of controls to allow blind screen reader users to turn off  or configure the live region announcements.  These controls would not affect the updates themselves.
